http/src/main/scala/org/http4s/blaze/http/http2/StreamState.scala

```diff
@@ -269,15 +292,32 @@ private abstract class StreamState(session: SessionCore)
         // Gotta make sure both sides agree that this stream is closed
         case None if !(sentEndStream && receivedEndStream) => Some(CANCEL.rst(streamId))
         case None => None
-        case Some(t: Http2Exception) => Some(t)
+        case ex@Some(_: Http2Exception) => ex
         case Some(EOF) => None
         case Some(other) =>
           logger.warn(other)(s"Unknown error in stream($streamId)")
           Some(INTERNAL_ERROR.rst(streamId, "Unhandled error in stream pipeline"))
       }
 
-      // Remove ourselves from the streamManager
-      session.streamManager.streamFinished(this, http2Ex)
+      // Remove ourselves from the streamManager. If the stream didn't exist, it is
+      // because this tream was reset by the remote, and thus we shouldn't send a
+      // RST ourselves. https://tools.ietf.org/html/rfc7540#section-5.4.2
+      val sendRst = session.streamManager.streamClosed(this)
+
+      http2Ex match {
+        case Some(ex: Http2StreamException) if sendRst =>
+          logger.debug(ex)(s"Sending stream ($streamId) RST")
+          val frame = session.http2Encoder.rstFrame(streamId, ex.code)
+          session.writeController.write(frame)
+          ()
+
+        case Some(ex: Http2SessionException) =>
+          logger.info(s"Stream($streamId) finished with session exception")
+          // TODO: this doens't feel like the responsibility of the StreamState
+          session.invokeShutdownWithError(http2Ex, "streamFinished")
+
+        case _ => () // nop
```

Comment:
I'd make this a `case None` to preserve exhaustiveness checking in case `http2Ex` ever gets widened.  I don't expect it to, but empty exception handlers in long methods are playing with fire when you refactor.

Reply:
Great idea, I hadn't considered that.